### PR TITLE
As a user I should be able to login so I can use my Target MVD account

### PR DESCRIPTION
**Pull request by keinydev**
[Trello Card](https://trello.com/c/NIB1cfPC/14-3-as-a-user-i-should-be-able-to-login-so-i-can-use-my-target-mvd-account)

### CODE REVIEW

File: spec/factories/users.rb [hunk 1]
```diff
@@ -18,7 +18,11 @@
   factory :user do
     email                  { Faker::Internet.unique.email }
     password               { Faker::Internet.password(min_length: 8) } 
-    gender                 { Faker::Internet.gender } 
+    gender                 { Faker::Gender.binary_type } 
     uid                    { email }
+
+    factory :user_without_uid do
```
Comment (antoalconti) [inline spec/factories/users.rb:24]: Great job!!  In this case, you can call the factory and set the param in nil or '', something like  create(:user, uid: '') in the spec

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -17,5 +17,16 @@
 require 'rails_helper'
 
 RSpec.describe User, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  describe 'callbacks' do 
+    context 'validation before saving the user' do
+      
+      user = FactoryBot.build(:user_without_uid)
+      
+      it 'should set uid as email' do
```
Comment (santir489) [inline spec/models/user_spec.rb:25]: `it 'sets uid as email' do`

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -17,5 +17,16 @@
 require 'rails_helper'
 
 RSpec.describe User, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  describe 'callbacks' do 
+    context 'validation before saving the user' do
+      
+      user = FactoryBot.build(:user_without_uid)
```
Comment (santir489) [inline spec/models/user_spec.rb:23]: Lets try to use `let` variables in specs
https://pololu.github.io/rpicsim/file.IntroductionToRSpec.html#label-Let+variables

Also, i think you can do
`let!(:user) { create(:user_without_uid) } `

**Review by santir489 — COMMENTED** [review]
Great! Left some comments

**Review by sebastiancaraballo — COMMENTED** [review]
Please remember to add the Trello card link to this PR description 💪

File: spec/factories/users.rb (hunk 1 above)
Comment (keinydev) [reply spec/factories/users.rb:24]: Thanks!

File: spec/models/user_spec.rb (hunk 3 above)
Comment (keinydev) [reply spec/models/user_spec.rb:23]: Thanks for the suggestion.

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -17,5 +17,16 @@
 require 'rails_helper'
 
 RSpec.describe User, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  describe 'callbacks' do 
+    context 'validation before saving the user' do
+      
+      let!(:user) { FactoryBot.build(:user, uid: '') }
```
Comment (santir489) [inline spec/models/user_spec.rb:23]: what about 
`let!(:user) { create(:user, uid: '') }` ? 
IMHO is more readable

**Review by santir489 — APPROVED** [review]
Good work! Left a minnor comment, it would be great if you could check it. 
Approved!

File: spec/requests/api/v1/sessions/create_spec.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  describe 'POST Create' do
+    context 'when login params is valid' do
+      before do
+        post @sign_in_url, params: @sign_in_params, as: :json
+      end
+      it 'returns status 200' do
+        expect(response).to have_http_status(200)
+      end
```
Comment (sebastiancaraballo) [inline spec/requests/api/v1/sessions/create_spec.rb:22]: Leave a blank line after each example.

File: spec/requests/api/v1/sessions/create_spec.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
+require "rails_helper"
+
+RSpec.describe "POST api/v1/users/sign_in", type: :request do
+  before(:each) do
+    @user = FactoryBot.create(:user)
```
Comment (sebastiancaraballo) [inline spec/requests/api/v1/sessions/create_spec.rb:5]:
```suggestion
    let(:user) { create(:user) }
```

File: spec/requests/api/v1/sessions/create_spec.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
+require "rails_helper"
+
+RSpec.describe "POST api/v1/users/sign_in", type: :request do
+  before(:each) do
+    @user = FactoryBot.create(:user) 
+    @sign_in_url = '/api/v1/users/sign_in' 
+    @sign_in_params = {
```
Comment (sebastiancaraballo) [inline spec/requests/api/v1/sessions/create_spec.rb:7]:
```suggestion
    let(:params) do
      ...
    end
```

File: spec/requests/api/v1/sessions/create_spec.rb [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        email: @user.email,
+        password: @user.password
+      }
+    }
+  end
+
+  describe 'POST Create' do
+    context 'when login params is valid' do
+      before do
+        post @sign_in_url, params: @sign_in_params, as: :json
```
Comment (sebastiancaraballo) [inline spec/requests/api/v1/sessions/create_spec.rb:18]: Try extracting this into a `subject` variable above.
`subject(:signin) { post @sign_in_url, params: params, as: :json }`

Also, replace @sign_in_url with its corresponding coming from running this command: `rails routes`

**Review by sebastiancaraballo — COMMENTED** [review]
Nice work. Please, add documentation by creating an `apiary.apib` file at the project root. Take a look at [this](https://apiblueprint.org/).

**Review by santir489 — APPROVED** [review]
🚀

**Review by sebastiancaraballo — APPROVED** [review]

**Pull request merged**